pkg/util/ip_reservation_test.go

```diff
@@ -0,0 +1,183 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"net"
+	"testing"
+)
+
+func initTestIPAllocator() *IPAllocator {
+	reserved := make(map[string]bool)
+	return &IPAllocator{
+		reserved: reserved,
+	}
+}
+
+func TestValidateCIDR(t *testing.T) {
+	testCIDR := "192.168.92.0/29"
+	ipAllocator := initTestIPAllocator()
+	err := ipAllocator.ValidateCIDR(testCIDR)
+	if err != nil {
+		t.Errorf(err.Error())
+	}
+
+	testCIDR = "192.168.92.0"
+	err = ipAllocator.ValidateCIDR(testCIDR)
+	if err == nil {
+		t.Errorf("Invalid CIDR %s evaluated to be valid", testCIDR)
+	}
+}
+
+func TestGetUnReservedIP(t *testing.T) {
+	ips := [8]string{"192.168.92.3/29", "192.168.92.10/29", "192.168.92.20/29", "192.168.92.28/29"} // Using IPs which are not the beginning IPs of /29 CIDRs to evaluate the edge case
+	cidr := "192.168.92.0/27"
+	// Reserved IPV4 ranges listed by the cloud provider
+	reservedIPV4Ranges := make(map[string]bool)
+	cases := []struct {
+		name          string
+		expected      string
+		errorExpected bool
+	}{
+		{
+			name:          "0/4 /29 IP blocks used",
+			expected:      "192.168.92.0/29",
+			errorExpected: false,
+		}, {
+			name:          "1/4 /29 IP blocks used",
+			expected:      "192.168.92.8/29",
+			errorExpected: false,
+		},
+		{
+			name:          "2/4 /29 IP blocks used",
+			expected:      "192.168.92.16/29",
+			errorExpected: false,
+		},
+		{
+			name:          "3/4 /29 IP blocks used",
+			expected:      "192.168.92.24/29",
+			errorExpected: false,
+		},
+		{
+			name:          "All /29 IP blocks used",
+			expected:      "",
+			errorExpected: true,
+		},
+	}
+
+	for index, test := range cases {
+		ipAllocator := initTestIPAllocator()
+		for i := 0; i < index; i++ {
+			// Distributing IPs between the reserved lists at the cloud provider end and the reserved list in the ipAllocator
```

Comment:
this setup should probably be a part of each test case definition

Reply:
Other functions don't accept the reserved list so created only for this.